scipy/stats/_distn_infrastructure.py

```diff
@@ -3208,38 +3197,55 @@ def fun(x):
         else:
             invfac = 1.0
 
-        tot = 0.0
-        low, upp = self._ppf(0.001, *args), self._ppf(0.999, *args)
-        low = max(min(-suppnmin, low), lb)
-        upp = min(max(suppnmin, upp), ub)
-        supp = np.arange(low, upp+1, self.inc)  # check limits
-        tot = np.sum(fun(supp))
-        diff = 1e100
-        pos = upp + self.inc
-        count = 0
+        x0 = self.ppf(0.5, *args, loc=loc)
+        if x0 < lb:
+            x0 = lb
+        if x0 > ub:
+            x0 = ub
 
-        # handle cases with infinite support
-
-        while (pos <= ub) and (diff > self.moment_tol) and count <= maxcount:
-            diff = fun(pos)
-            tot += diff
-            pos += self.inc
-            count += 1
+        # XXX: increase chunksize. Requires fixing skellam.pmf for large x.
 
-        if self.a < 0:  # handle case when self.a = -inf
-            diff = 1e100
-            pos = low - self.inc
-            while ((pos >= lb) and (diff > self.moment_tol) and
-                   count <= maxcount):
-                diff = fun(pos)
-                tot += diff
-                pos -= self.inc
-                count += 1
-        if count > maxcount:
-            warnings.warn('expect(): sum did not converge', RuntimeWarning)
+        count, tot = 0, 0.
+        for x in _iter_chunked(x0, lb, ub, chunksize=chunksize, inc=self.inc):
+            count += x.size
+            delta = np.sum(fun(x))
+            tot += delta
+            if abs(delta) < tolerance * x.size:
+                break
+            if count > maxcount:
+                warnings.warn('expect(): sum did not converge', RuntimeWarning)
+                break
         return tot/invfac
 
 
+def _iter_chunked(x0, a, b, chunksize=4, inc=1):
+    """Iterate over a support of a distribution in chunks.
+
+    Starts at `x0` (ppf(0.5) could be a good starting point).
+    Generates all integers with step inc from the closed interval [a, b]
+    unless either of `a` or `b` is infinite.
+    Handles all cases with the support being finite, infinite or bi-infinite.
+
+    """
+    chunksize *= inc
+
+    # short-circuit if possible
+    if (b - a) <= chunksize:
+        yield np.arange(a, b+1, inc)
+        return
+
+    # OK, iterate. Expand in both directions all the way to the bounds [a, b].
+    lo, hi = x0, x0
+    while (lo > a) or (hi < b):
+        delta = min(chunksize, b - hi)
+        res_p = np.arange(hi, hi + delta + 1, inc)
+        hi += delta
+        delta = min(chunksize, lo - a)
+        res_n = np.arange(lo - delta, lo, inc)
+        lo -= delta
```

Comment:
Maybe not very relevant for the current distributions, except skellam:
Going up and down at the same time doesn't allow stopping early in one of the directions if all probabilities there are "negligible", but not in the other direction.